omod/src/main/java/org/openmrs/module/smartonfhir/web/servlet/SmartLaunchOptionSelected.java

```diff
@@ -30,17 +32,42 @@
 import org.keycloak.jose.jws.JWSBuilder;
 import org.keycloak.representations.JsonWebToken;
 import org.openmrs.module.smartonfhir.util.SmartSecretKeyHolder;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.util.UriComponentsBuilder;
 
+@Slf4j
 public class SmartLaunchOptionSelected extends HttpServlet {
 	
 	public void doGet(HttpServletRequest req, HttpServletResponse res) throws IOException {
 		String token = getParameter(req, "token");
 		String patientId = getParameter(req, "patientId");
 		String visitId = getParameter(req, "visitId");
 		String decodedUrl = URLDecoder.decode(token, StandardCharsets.UTF_8.name());
-		String launchType = getParameterFromStringUrl(decodedUrl, "launchType");
 		
-		if (launchType.equals("encounter") && visitId == null) {
+		String jwtKeyToken = null;
+		try {
+			jwtKeyToken = getParameterFromStringUrl(decodedUrl, "key");
+		}
+		catch (URISyntaxException e) {
+			log.error("Verification exception while trying to determine launchType", e);
+			return;
+		}
+
+		String launchTypeString = null;
+
+		try {
+			launchTypeString = getLaunchTypeString(jwtKeyToken);
+		} catch (VerificationException e) {
+			log.error("Error while extracting the launch type from token");
```

Comment:
This way we get the exception logged server-side. Useful for debugging.

```suggestion
			log.error("Error while extracting the launch type from token", e);
```